Update Game_mac build for Xcode 6

nlap opened the pull request:
Here are some improvements to xcode project in case people need to modify it

* Updates Xcode build settings to allow Game.app to be built on Xcode 6+
* Adds sandbox support to allow games to be submitted to Mac App Store
* Removes build folder that doesn't need to be in source control

Remaining issues:

* App crashes in full screen due to OS X 10.9+ changes. I think this needs to be fixed in OgreOSXCocoaWindow.mm
* App segfaults if closed using window's red X button instead of through game menu. Not sure how to fix yet

Orvid commented:
Do these changes prevent building with previous versions of Xcode?

Orvid commented on Project/Src/Game_Mac/English.lproj/MainMenu.xib:
```diff
@@ -676,8 +675,9 @@
-				<string key="NSMaxSize">{3.40282e+38, 3.40282e+38}</string>
+				<string key="NSScreenRect">{{0, 0}, {1920, 1177}}</string>
+				<string key="NSMaxSize">{10000000000000, 10000000000000}</string>
+				<bool key="NSWindowIsRestorable">YES</bool>
```
The default window size here is too large for Mac's without a retina display as well as smaller Macbook Air's. What is the behavior of `NSScreenRect` on those platforms where it's larger than the actual screen size?

Also, assuming `NSWindowIsRestorable` refers to the same thing as on windows, should the window really be resizable?

nlap replied:
This is a good catch. I did this on a retina mac and forgot about smaller screens ;)

The NSWindowIsRestorable is new to OS X since this project was made. I'm not sure what this will mean for this project.

Orvid commented on Project/Src/Game_Mac/English.lproj/MainMenu.xib:
```diff
@@ -2343,40 +1789,20 @@
-			<string key="NS.key.0">com.apple.InterfaceBuilder.CocoaPlugin.macosx</string>
-			<integer value="1050" key="NS.object.0"/>
-		</object>
+		<bool key="IBDocument.previouslyAttemptedUpgradeToXcode5">NO</bool>
```
Should this be here? I this is upgrading to XCode 6.

nlap replied:
This was added automatically by xcode 6 upgrade

Orvid commented on Project/Src/Game_Mac/Game.xcodeproj/project.pbxproj:
```diff
@@ -227,8 +229,20 @@
+			};
 			buildConfigurationList = C01FCF4E08A954540054247B /* Build configuration list for PBXProject "Game" */;
-			compatibilityVersion = "Xcode 3.1";
+			compatibilityVersion = "Xcode 3.2";
```
Worth making a note of?

nlap replied:
I'm not sure. Xcode upgrade added this.

Orvid commented on Project/Src/Game_Mac/Game.xcodeproj/project.pbxproj:
```diff
@@ -347,20 +361,26 @@
 		18DA27FA13351DFE00B031E4 /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
+				CONFIGURATION_BUILD_DIR = build/Debug;
```
Where did previous versions build to? Could this cause issues with any automated build scripts?

nlap replied:
Previous version built to /Temp. This isn't a valid temp directory on OS X anymore, so Xcode doesn't have permission to build here without root. Shouldn't cause problems with build scripts. This is the standard directory structure for xcode builds now.

Orvid commented on Project/Src/Game_Mac/Game.xcodeproj/project.pbxproj:
```diff
@@ -368,38 +388,55 @@
 		18DA282113351EEA00B031E4 /* Debug */ = {
 			isa = XCBuildConfiguration;
 			buildSettings = {
+				COMBINE_HIDPI_IMAGES = YES;
```
The previous behavior was what, and does this have any implications for rendering or filesize?

nlap replied:
Xcode upgrade added this. I think this is new behavior as retina Macs didn't exist when this project was made. It just combines any 1x and 2x artwork into one TIFF file to save space. Currently it has no affect on the build, since there's not artwork in the project.

Orvid commented on Project/Src/Game_Mac/Game.xcodeproj/project.pbxproj:
```diff
@@ -368,38 +388,55 @@
+				COMBINE_HIDPI_IMAGES = YES;
+				CONFIGURATION_BUILD_DIR = build/Debug;
+				CONFIGURATION_TEMP_DIR = build/Game.app/Debug;
+				FRAMEWORK_SEARCH_PATHS = "$(inherited)";
```
Do we no longer need to link against Cg, or is it part of the normally inherited search paths now?

nlap replied:
Cg is still needed, but it's added to Game.app by the build 'run script' phase. No need to have it in the framework search path.

Orvid commented on Project/Src/Game_Mac/Game.xcodeproj/project.pbxproj:
```diff
@@ -368,38 +388,55 @@
-					"$(inherited)",
-					"\"$(SRCROOT)/../../../Engine/Src/Core/External/Cg/lib/MacOSX\"",
-				);
+				CODE_SIGN_ENTITLEMENTS = Game.entitlements;
```
I understand the reason why the entitlements are needed, but want to make sure this doesn't try to sign the output assembly. This does not sign the output assembly, correct?

nlap replied:
Correct. This just adds the file to the code signing list, in case a signing key is added to the project. Output is not signed currently.

Orvid commented on Project/Src/Game_Mac/Game.xcodeproj/project.pbxproj:
```diff
@@ -368,38 +388,55 @@
+				CONFIGURATION_TEMP_DIR = build/Game.app/Release;
+				FRAMEWORK_SEARCH_PATHS = "$(inherited)";
 				INFOPLIST_FILE = "Game-Info.plist";
+				ONLY_ACTIVE_ARCH = NO;
```
Any particular reason this got added?

nlap replied:
Automatic upgrade by xcode. I think this prevents 64-bit from building (only 32-bit supported for Mono to work anyway).

Orvid commented on Project/Src/Game_Mac/Game.xcodeproj/project.pbxproj:
```diff
@@ -407,14 +444,16 @@
 				GCC_C_LANGUAGE_STANDARD = gnu99;
 				GCC_OPTIMIZATION_LEVEL = 0;
 				GCC_WARN_ABOUT_RETURN_TYPE = YES;
+				MACOSX_DEPLOYMENT_TARGET = 10.7;
```
Does this prevent the app from running on earlier versions of OSX?

nlap replied:
Yes, this will be necessary to support full screen games. 10.6 does not have official fullscreen mode. More changes are needed in OGRE part of NeoAxis I think to support this properly.

Latest stats have 97%+ of Macs are running 10.7 or later. https://www.gosquared.com/global/mac/yosemite

Orvid commented:
For future reference, it would be preferred if each of those larger changes went in their own commit, as it makes it significantly easier to review.

nlap commented:
Thanks for the review. Noted that there should be more commits in future. Many of these changes were done automatically in one step by Xcode.

EDIT: yes, I think this will prevent previous versions of xcode from opening the project.

Orvid commented:
Assuming the behavior of `NSScreenRect` and `NSWindowIsRestorable` is acceptable, this looks good to merge.

nlap commented:
Yes, it looks like it will be fine.

NSScreenRect is "The screen size on which the view was encoded(used for scaling)". So my screen size. The actual Game.app window size didn't change since the last version.

NSWindowIsRestorable allows window state and size to be preserved between sessions. I think this is fine since the game engine will resize the view according to configuration anyway.

Orvid commented:
Then, assuming @IvanEfimov is ok with this change, we can merge it. (I'll let him click the merge button though)

IvanEfimov commented:
Hi!
Thank you for first pull! Cool!

Hmmm, want to save support of old XCode versions.
Fullscreen problem on OS X 10.9+ is already fixed. NeoAxis 3.1 coming soon.

nlap commented:
Thanks for fixing fullscreen!

Problem is that app doesn't build at all on new Xcode 6 (maybe 5 too). Since xcode upgrades are free, I think it should support latest version?

Orvid commented:
I suppose it would depend on if you can install XCode 6 on OSX 10.7?

How complex a process is it to perform the project upgrade? NeoAxis has for a long time required users of newer versions of Visual Studio to first upgrade the project before being able to build it, but, as that's a quick and painless process, it's never really been an issue.

Even if it is a moderately painless process, it might be worth creating a second project for XCode 6, as the project files themselves are rarely touched, so I don't believe keeping them in-sync should be much of an issue.

nlap commented:
Xcode 6 only works on 10.10. 10.7 only supports Xcode 4.

The upgrade process is trickier than Visual Studio for this project. There is an automated step, but in addition several build settings need to be changed, and the build script needed to be rewritten by hand.

Orvid commented:
That being the case, what would @IvanEfimov think about a separate project for XCode 6? (knowing that the project will almost never change, so it is unlikely to need to be updated)
